Adds capability for NINO34 index and spectra plots

vanroekel opened the pull request:
This adds NINO34 functionality.  It computes the NINO34 index following the standard algorithm (e.g. Trenberth 1997) it also computes the power spectra, along with the 95 and 99% confidence intervals (using a chi-squared test).  A red noise best-fit is also included.  This algorithm follows closely what is used in NCL.  

It has been tested on a g-case on anvil and beta0 and beta1 on edison.

vanroekel commented:
Here are some beta0 plots, from the coupled low-res on edison.  Config file is /global/homes/l/lvroekel/MPAS-Analysis/config.20161117.beta0.A_WCYCL1850.ne30_oEC.edison

![nino34_20161117 beta0 a_wcycl1850 ne30_oec edison](https://cloud.githubusercontent.com/assets/13662545/24157814/1a008c10-0e20-11e7-9632-143e07c212a9.png)

The spectra plot is below

![nino34_spectra_20161117 beta0 a_wcycl1850s ne30_oec_icg edison_longtime](https://cloud.githubusercontent.com/assets/13662545/24157892/51722302-0e20-11e7-86b6-997e55e38896.png)

vanroekel commented:
@xylar @pwolfram and @milenaveneziani I think I have an issue in the spectral computation, please do not review that portion just yet.  The time series and plotting capabilities should be fine though.

pwolfram commented:
@vanroekel, sounds good.  Please let me know when you are ready for a comprehensive review.  Also, do you think it could be helpful to add a legend / labels to the plots?  I'd recommend a legend for the power spectra plot at the least.

pwolfram commented:
Once you've done this and fixed the spectra issue can you please squash?  I'll give you some comments outside the spectra computation in the mean time.

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+
+def _autocorr(x, t=1):
+    """
+    Computes lag one auto-correlation for the NINO34 spectra calculation
```
If `t \=1` this would no longer be a "lag one auto-correlation", correct?  We may want to have a clearer docstring that clarifies this point.

pwolfram replied:
This is basically just a lagged-autocorrelation, right?

vanroekel replied:
yes, exactly, I can add the bit about t/=1

vanroekel replied:
done

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    rsq = r**2
+
+#   In the temp2 variable, f is converted to give wavenumber i.e., 0,1,2,...,N/2
+    temp2 = r2*np.cos(2.*np.pi*f*86400.*30.)
```
Magic numbers?

vanroekel replied:
done

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    Last Modified: 03/20/2017
+    """
+
+    return pd.Series.rolling(inputData, 5, center=True, min_periods=1).mean()
```
Why not just use the xarray rolling mean?  This would avoid the pandas dependency.

pwolfram replied:
We are also missing the following line at the end of the file: `# vim: foldmethod=marker ai ts=4 sts=4 et sw=4 ft=python`

vanroekel replied:
was copying code from the plot_timeseries routine.

milenaveneziani replied:
yes, we are using the panda rolling mean everywhere, I think.
@vanroekel: could you please change '5-point' average with '5-month' average in the few places where it is mentioned? (or something along that line)

vanroekel replied:
changed to xarray.rolling

vanroekel replied:
changed 5-point to 5-month

vanroekel replied:
@xylar and @pwolfram when I tried switching to xr.rolling I get the following error
```
Traceback (most recent call last):
  File "./run_analysis.py", line 260, in <module>
    analysis(config)
  File "./run_analysis.py", line 158, in analysis
    variableMap=oceanVariableMap)
  File "/global/u2/l/lvroekel/MPAS-Analysis/mpas_analysis/ocean/nino34_index.py", line 99, in nino34_index
    nino34 = compute_nino34_index(SSTregions[:, regionIndex], config)
  File "/global/u2/l/lvroekel/MPAS-Analysis/mpas_analysis/ocean/nino34_index.py", line 150, in compute_nino34_index
    return _running_mean(anomalySST)
  File "/global/u2/l/lvroekel/MPAS-Analysis/mpas_analysis/ocean/nino34_index.py", line 225, in _running_mean
    return inputData.rolling(Time=5, center=True, min_periods=1).mean()
  File "/global/project/projectdirs/acme/software/anaconda_envs/edison/lib/python2.7/site-packages/xarray/core/common.py", line 69, in wrapped_func
    'Rolling window operation does not work with dask arrays')
NotImplementedError: Rolling window operation does not work with dark arrays

```
This seems to be a known issue: https://github.com/pydata/xarray/issues/1279  Are there objections to retaining pandas.rolling for now?

xylar replied:
Foiled again!

I have no objection.  Could you convert back to an xarray dataset right away (in `_running_mean`) instead of in the plotting routine?  That will mean minimal confusion with what is a pandas data set and what is xarray.

vanroekel replied:
Okay, can do.

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    The algorithm follows the NCL cvdp package see
+    http://www.cesm.ucar.edu/working_groups/CVC/cvdp/code.html
+
+    NOTE: this routine requires the signal and stats module from scipy
```
Should you just import these directly in this function then if they are only used here, which it appears is the case?  This would be clearer and avoid need for this note.

vanroekel replied:
Okay, I had this originally.  Will return the imports here.

vanroekel replied:
done

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    return np.corrcoef(np.array([x[0:len(x)-t], x[t:len(x)]]))
+
+
+def _running_mean(inputData):
```
Nice, makes the code cleaner to have this function to avoid duplicated code.

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
 from ..constants import constants
 
+
+def nino34_spectra_plot(config, f, ninoSpectra, confidence95, confidence99,
```
I'd consider having default `xLabel` and `yLabel` arguments for enhanced clarity.

vanroekel replied:
done

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+        plt.close()
+
+
+def nino34_timeseries_plot(config, nino34Index, title, xlabel, ylabel,
```
I'd consider having default `xLabel` and `yLabel` arguments.

vanroekel replied:
done

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+    Last Modified: 03/21/2017
+    """
+
+    plt.plot(f[2:-3], ninoSpectra[2:-3], 'k', linewidth=linewidths)
```
An idea-- maybe just write something like this so that you can easily change the limits later if needed:
```python
# comment on why you are plotting the [2:-3] slice
def plot_slice(x, y, *wargs, slice=slice(2,-3),  **kwargs):
    plt.plot(x[slice], y[slice], *wargs, **kwargs)

plot_slice(f, ninoSpectra, 'k', linewidth=linewidths)
...
```

vanroekel replied:
given that the 5 point running mean is specified in the nino 34 algorithm I think it is okay as is.

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+                           maxXTicks=20):
+    """
+    Plots the nino34 time series and power spectra in an image file
+     Parameters
```
There is a spurious space in front of "Parameters"

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+
+    nino34Index = xr.DataArray.from_series(nino34Index)
+    y1 = nino34Index[2:-3]
+    nt = np.size(nino34Index[2:-3])
```
Cache time too for use in plotting below:
```python
time = nino34Index.Time[2:-3].values
```

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+    plt.plot(nino34Index.Time[2:-3].values, 0.4*np.ones(nt), '--k',
+	     linewidth=linewidths)
+    plt.plot(nino34Index.Time[2:-3].values, -0.4*np.ones(nt), '--k',
+             linewidth=linewidths)
```
Should you annotate these regions with words on the plot to clarify?

vanroekel replied:
Do you mean something like red=El Nino, blue = La Nina?

xylar replied:
I'm guessing he means more like saying something about what the upper and lower lines mean.  I think that would belong in a figure caption, rather than on the figure itself.  Something like "index values outside the dashed lines indicate significant El Nino (positive) or La Nina (negative) events".  I don't think that belongs in the plot itself.

pwolfram commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+                major.append(date_to_days(year=year, month=month,
+                                          calendar=calendar))
+        formatterFun = partial(_date_tick, calendar=calendar,
+                               includeMonth=True)
```
Seems like we should have some codified code to handle ticks in a shared module for this if/else statement above.  It would be better if this code were a clear, reusable function if possible.

pwolfram replied:
Probably would include code through `plt.autoscale(enable=True, axis='x', tight=True)` below in this function.

vanroekel replied:
I think this is beyond the scope of this PR

xylar replied:
@vanroekel, I think @pwolfram is right and that it *is* in the scope of this PR.  If you're the first to reuse a bit of code from another plotting routine, I think it's your responsibility to make a helper function (say `_format_year_month_ticks()`) and use the same shared code in both.  Otherwise, if we want to change the how ticks work (which actually seems very likely), we have to do it in 2 places.

xylar replied:
I think the shared routine would include all of the code starting at:
```
ax = plt.gca()
```
and ending below at
```
plt.autoscale(enable=True, axis='x', tight=True)
```

vanroekel replied:
okay, I guess I can try fix that

vanroekel replied:
@xylar I've made a stab at the helper function.  Could you please review (in particular the doc-string)?  It seems to work for me.

xylar replied:
So, just a few things I'd change.  First, `plt` is not something that should be passed as an argument because it's a module.

Here's a slightly modified docstring:
```
    Formats tick labels and positions along the x-axis for time series / index 
    plots

    Parameters
    ----------
    calendar : {'gregorian', 'gregorian_noleap'}
        the calendar of the data set being plotted

    minDays : float
        start time for labels

    maxDays : float
        end time for labels

    Author
    ------
    Xylar Asay-Davis

    Last Modified
    -------------
    03/22/2017
```

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    """
+    Computes power spectra of Nino34 index.
+
+    nino34Index is the NINO index computed by compute_nino34_index
```
Perhaps consider briefly summarizing equations for how this should be calculated, e.g., rationale for computation.

vanroekel replied:
This is just an FFT, what is there to add here?

pwolfram replied:
This is what I meant, something like "NINO3.4 is the average sea surface temperature anomaly in the region bounded by 5°N to 5°S, from 170°W to 120°W. " http://www.eldersweather.com.au/climimage.jsp?i=nino34

vanroekel replied:
This is already in the nino34 index doc string, would you like it here as well?

pwolfram replied:
Thanks, you are right @vanroekel, we don't need it here.  I would just add a reference to the other function in the doc string for completeness.

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    namelist = NameList(namelistFileName, path=inDirectory)
+    calendar = namelist.get('config_calendar_type')
+
+    # Compute monthly average climatology of SST
```
`# Compute monthly average and anomaly of climatology of SST` and probably remove space between two lines of computation.

vanroekel replied:
done

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    date_to_days, days_to_datetime
+
+
+def nino34_timeseries(config, streamMap=None, variableMap=None, SSTregions=None):
```
Please wrap with `# {{{` and `return # }}}` for folds in all functions in this file.

milenaveneziani replied:
Why is SSTregions passed through the function call?

vanroekel replied:
SSTregions shouldn't be there.  Removing.

pwolfram commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+from ..shared.plot.plotting import nino34_timeseries_plot, nino34_spectra_plot
```
Do you want to have a module header docstring block? For example, this would be something like
```python
"""
Computes NINO34 diagnostic.

Author: Luke Van Roekel
Last Modified: 03/21/2017
"""
```

milenaveneziani replied:
would call it 'Nino3.4 index diagnostics'

vanroekel replied:
done.

vanroekel commented:
@pwolfram thanks for the comments.  I will work on these.  Agreed also on spectra testing.  A question, do you have any python code that generates a red noise time series?  It would be helpful for testing the code.

pwolfram commented:
@vanroekel, I don't have code of my own but you can try http://relay.actimar.fr/~raynaud/spanlib/_modules/spanlib/analyzer.html#RedNoise and https://github.com/python-acoustics/python-acoustics/blob/master/acoustics/generator.py to get different colors of noise.

pwolfram commented:
You can also use this page: http://www.redblobgames.com/articles/noise/introduction.html and 8.1 to get the explicit formulas, e.g., 
![screenshot 2017-03-21 11 34 54](https://cloud.githubusercontent.com/assets/4295853/24161604/85428564-0e2a-11e7-9c1e-b2dbed755e32.png)

milenaveneziani commented on config.default:
```diff
@@ -286,11 +286,7 @@ movingAveragePoints = 12
-# output, movingAveragePoints=12 corresponds to a 12-month moving average
-# window)
-movingAveragePoints = 12
+regionIndex = 5
```
@vanroekel: I would personally change the section name to something like `Nino34Index`, since you are adding spectra.
To echo @pwolfram comment: don't you need the movingAveragePoints, only different from 12 (something like a 5-month average, if I remember correctly)? If you would like to change the `regionIndex` name, I would probably do it for all scripts consistently.

vanroekel replied:
@milenaveneziani and @pwolfram I don't need the movingAverage.  I didn't see the reason to retain.  The nino3.4 algorithm specifies a 5-month average so it seemed unnecessary and unwise to have the possibility for the user to pick something else.

@milenaveneziani I don't understand the second comment about regionIndex.  Are you saying to refer to regionIndicesToPlot as before?  I can change that.

milenaveneziani replied:
I agree about the moving average.
And yes, I was referring to regionIndex name. I don't mind changing it, but we should do use the same everywhere (I think it would only be for sst and ohc timeseries).

vanroekel replied:
@milenaveneziani, my only hesitation about the rename is that I think it would get missed in an option like 'all_timeSeries' and would be the odd analysis out.  But I may be may off base here

milenaveneziani replied:
Do you mean the rename of the config section?
I see your point. I have the same for the MOC. I think it should be OK as long as we add instructions in the config.default. But I will let @xylar comment on this too.

vanroekel replied:
I think I see a way to move around the rename.  I'm testing now.  

On a more general point.  Given that this routine is only intended to plot nino3.4, which is region 5, I think we could just get rid of the entire timeSeriesNino3.4 section.  Or would it be useful to retain the option for nino 3 and or nino 4?  I'm fine with either.  I would be interested in thoughts from all ( @xylar, @pwolfram, and @milenaveneziani )

milenaveneziani replied:
I would keep it: we may want to add/modify something in the future.

milenaveneziani replied:
@xylar, @vanroekel, about regions, region names/indexes: at some point, we will have to update a few AM's (including the `surfaceAverages` used here, and `layerWeightedAverages` used for the OHC plot) to include better handling of ocean basin and regional calculations, using maskfiles instead of the hardwired regions that we have now. When that time comes (right now this is not a huge priority), we can get regionNames and regionGroupNames from the mask files, or maybe it could be part of the AM output.
About removing `compareWithObservations`: we probably need to do a bit of cleanup in general, although not in this PR, also including `compareWithModel`.

milenaveneziani commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    their mpas_analysis counterparts.
+
+    If present, variableMap is a dictionary of MPAS-O variable names that map
+    to their mpas_analysis counterparts.
```
would you mind adapting this doc string to the standard format we have been trying to adhere to in the last few PRs? You can use the MOC script as an example. thanks.

milenaveneziani commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    f = 1.0 / (1E-9 + f*86400.*365)
+    print ' Plot NINO3.4 index and spectra...'
+    xLabel = 'Time [years]'
+    yLabel = '[$^\circ$ C]'
```
@pwolfram: I think what is currently here is how we do it now for other plotting routines.

vanroekel replied:
I've moved these to defaults in the plot function as I don't foresee these changing once in place.

milenaveneziani commented:
Very nice, thanks @vanroekel. I have added some comments here and there, but I think this is already in pretty good shape.
About the plots themselves, some low level comments are: what about removing the extra space between the degree sign and C in the timeseries plot?, and maybe reduce the linewidths a bit in the spectrum plot?

xylar commented:
@vanroekel, I'm looking forward to giving this a thorough look and trying it out tomorrow.

vanroekel commented:
The spectral capability is fixed.  I'll try and get to the requests here before @xylar dives in.

vanroekel commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    xLabel = 'Time [years]'
+    yLabel = '[$^\circ$ C]'
+
+    figureName = '{}/NINO34_{}.png'.format(plotsDirectory, mainRunName)
```
I'm unsure what you are requesting here.  Do you want me to change something in the figure name?

milenaveneziani commented:
@vanroekel: I ran on edison using beta1 data, which I think it's the same case you showed us plots above, but I get a somehow different result, even in the time series (I imagine the spectrum *should* be different). Do you understand why?
Here is what I get:

![nino34_20170313 beta1 a_wcycl1850s ne30_oecv3_icg edison](https://cloud.githubusercontent.com/assets/8237579/24181701/5daf6adc-0e82-11e7-9720-400edc4fb532.png)
![nino34_spectra_20170313 beta1 a_wcycl1850s ne30_oecv3_icg edison](https://cloud.githubusercontent.com/assets/8237579/24181700/5daf3044-0e82-11e7-9cd3-f0756934acd1.png)

milenaveneziani commented:
One other thing I forgot to mention: could you add a legend to the power spectrum plot mentioning what the different lines represent? maybe you can even do it in the title if the fontsize is reduced.

vanroekel commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    end = days_to_datetime(np.amax(ds.Time.max()), calendar=calendar)
+
+#   Convert frequencies to period in years
+    f = 1.0 / (1E-9 + f*86400.*365)
```
done

vanroekel commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    			   xLabel, yLabel, figureName, linewidths=3, calendar=calendar)
+
+    xLabel = 'Period (years)'
+    yLabel = 'Power Spectrum'
```
Done

vanroekel commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,198 @@
+    This routine requires regionSST to be the SSTs in the nino3.4 region ONLY.
+    It is defined as lat > -5S and lat < 5N and lon > 190E and lon < 240E.
+
+    Further, regionSST must be a dataArray
```
done

milenaveneziani commented:
Forget about what I said above.. You ran with beta0.

vanroekel commented:
@milenaveneziani one other difference with my plot above is that there were some errors in the spectral code that i have fixed.  So your plot should be different than mine.  My new plot from the fixed version is very similar to yours

![nino34_spectra_20170313 beta1 a_wcycl1850s ne30_oecv3_icg edison](https://cloud.githubusercontent.com/assets/13662545/24182181/97ab61ca-0e85-11e7-9e4b-eb430005ee62.png)

I'm working on the legend now.  I've also reduced the line thickness too.

vanroekel commented:
@xylar I am nearly done addressing comments from @pwolfram and @milenaveneziani   I want to make one more change that sets the spectra y-axis range in a more useful way, but won't get to that until the morning.  Feel free to add comments based on what is here now.  I don't want to push the updates yet as the xarray rolling mean is not behaving as expected yet.  If you'd rather wait, I will get to this first thing in the morning.

xylar commented on mpas_analysis/ocean/nino34_timeseries.py:
```diff
@@ -0,0 +1,209 @@
+
+    NOTE: this routine requires the signal and stats module from scipy
+
+    Author: Luke Van Roekel
```
@vanroekel, could you state clearly what the function returns (preferably following the convention of other newer docstrings):
```
Returns
-------
f : numpy.array
    blah blah

pxxSmooth : numpy.array
    blah blah
```

xylar commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+    plt.figure(figsize=figsize, dpi=dpi)
+
+    nino34Index = xr.DataArray.from_series(nino34Index)
+    y1 = nino34Index[2:-3]
```
I believe if you switch(ed) the running mean to use xarray,  `nino34Index` is now an `xarray.DataSet` object.  If so, you should simply be able to slice it once: `nino34Index = nino34Index[2:3]` and get rid of all the other slicing below:
```
y1 = nino34Index.values
nt = len(y1)
time = nino34Index.Time.values
minDays = time[0]
maxDays = time[-1]
```

xylar commented on run_analysis.py:
```diff
@@ -149,12 +149,13 @@ def analysis(config):  # {{{
-#         nino34_timeseries(config)
+    if checkGenerate(config, analysisName='timeSeriesNino34',
+                     mpasCore='ocean', analysisCategory='timeSeries'):
+         print ""
```
This is where it should be fine if you have `analysisName='indexNino34'` but still keep `analysisCategory='timeSeries'`.  That way, when you do `--generate all_timeSeries`, `indexNino34` will be included among the tasks that run.

xylar commented:
@vanroekel,  this is very nice work!

I added some comments you can address as you work out rolling mean.  I'll hold off on testing the code itself until you push some changes.

xylar commented:
@vanroekel, could you add `indexNino34` (or whatever we end up calling it) to the list of possible analysis tasks associated with the `generate` flag in `config.default`?

xylar commented:
Plotting a very short time series, it becomes clear that transitions between positive and negative index are not being plotted.

![nino34_gmpas-qu240](https://cloud.githubusercontent.com/assets/4179064/24194211/826deb96-0ef5-11e7-9ebc-f5a073f516be.png)

Is that an expected behavior?  I wonder if the plotting routine could be made to handle these transitions without too much trouble?  I'll look into it.

xylar commented:
@vanroekel, could you make sure that all the routines you've written adhere to the PEP8 conventions?  There are 2 common ways to do that.  You can run flake8 (which you can install with conda) or you can run the IDE spyder and turn on the PEP8 warnings under Tools > Preferences > Editor > Code Introspection/Analysis > code analysis (pyflake).  I far prefer the latter option because you get real-time feedback as you write code.

We've been tolerating longer lines than 79 characters in the specific case that it makes strings more readable and intuitive (e.g. multiling error messages) but we've been adhering pretty strictly to the other PEP8 conventions.

xylar commented:
The solution to the plotting issue from above turns out to be simply adding `interpolate=True` to the arguments to `fill_between`:
![nino34_gmpas-qu240](https://cloud.githubusercontent.com/assets/4179064/24194928/264741a2-0ef8-11e7-9be5-e5c5ee4e493d.png)
@vanroekel, could you add this?

vanroekel commented:
@xylar I've added a bit of code to automatically set the y-lim max in the spectra plot, the legend, and your interpolate-true.  I have yet to do your clean-up requests.  But I think this is ready for your testing.

vanroekel commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,187 @@
+        the title of the plot
+
+    xlabelTimeSeries, ylabelTimeSeries, xlabelSpectra, ylabelSpectra : str
+        axis labels
```
yes indeed.  I will clean this up.

milenaveneziani commented:
@xylar, @vanroekel, I had put the following in the review section, but it's not disappeared due to the new changes.
About regions, region names/indexes: at some point, we will have to update a few AM's (including the `surfaceAverages` used here, and `layerWeightedAverages` used for the OHC plot) to include better handling of ocean basin and regional calculations, using maskfiles instead of the hardwired regions that we have now. When that time comes (right now this is not a huge priority), we can get regionNames and regionGroupNames from the mask files, or maybe it could be part of the AM output.

About removing `compareWithObservations`: we probably need to do a bit of cleanup in general, although not in this PR, also including `compareWithModel`.

milenaveneziani commented:
@vanroekel: I have updated to the newer code for testing, but I can't see the main nino34 routine in mpas-analysis/ocean any longer..

vanroekel commented:
@xylar, sorry about that, I changed the routine names to be index based and
not timeSeries, and forgot the git add nino34_index.py prior to commit.
Try now.

On Wed, Mar 22, 2017 at 9:41 AM, Xylar Asay-Davis <notifications@github.com>
wrote:

> *@xylar* commented on this pull request.
> ------------------------------
>
> In mpas_analysis/shared/plot/plotting.py
> <https://github.com/MPAS-Dev/MPAS-Analysis/pull/146#discussion_r107449467>
> :
>
> > +    end = days_to_datetime(np.amax(maxDays), calendar=calendar)
> +
> +    if (end.year - start.year > maxXTicks/2):
> +        major = [date_to_days(year=year, calendar=calendar)
> +                 for year in np.arange(start.year, end.year+1)]
> +        formatterFun = partial(_date_tick, calendar=calendar,
> +                               includeMonth=False)
> +    else:
> +        # add ticks for months
> +        major = []
> +        for year in range(start.year, end.year+1):
> +            for month in range(1, 13):
> +                major.append(date_to_days(year=year, month=month,
> +                                          calendar=calendar))
> +        formatterFun = partial(_date_tick, calendar=calendar,
> +                               includeMonth=True)
>
> So, just a few things I'd change. First, plt is not something that should
> be passed as an argument because it's a module.
>
> Here's a slightly modified docstring:
>
>     Formats tick labels and positions along the x-axis for time series / index
>     plots
>
>     Parameters
>     ----------
>     calendar : {'gregorian', 'gregorian_noleap'}
>         the calendar of the data set being plotted
>
>     minDays : float
>         start time for labels
>
>     maxDays : float
>         end time for labels
>
>     Author
>     ------
>     Xylar Asay-Davis
>
>     Last Modified
>     -------------
>     03/22/2017
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MPAS-Dev/MPAS-Analysis/pull/146#discussion_r107449467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANB5UfpjQowQ925QkEC2N1HbAHZ0YeiEks5roUEagaJpZM4MkDce>
> .
>

xylar commented:
I get the following error:
```
 Computing NINO3.4 power spectra...
/home/xylar/anaconda2/envs/default/lib/python2.7/site-packages/numpy/lib/function_base.py:2569: RuntimeWarning: invalid value encountered in true_divide
  c /= stddev[:, None]
 Plot NINO3.4 index and spectra...
Traceback (most recent call last):
  File "./run_analysis.py", line 260, in <module>
    analysis(config)
  File "./run_analysis.py", line 158, in analysis
    variableMap=oceanVariableMap)
  File "/home/xylar/code/mpas-work/analysis/NINO34/mpas_analysis/ocean/nino34_index.py", line 119, in nino34_index
    'NINO3.4 power spectrum', figureName, linewidths=2)
  File "/home/xylar/code/mpas-work/analysis/NINO34/mpas_analysis/shared/plot/plotting.py", line 106, in nino34_spectra_plot
    maxValueNinoSpectra = ninoSpectra[minIndex:maxIndex].values.max()
  File "/home/xylar/anaconda2/envs/default/lib/python2.7/site-packages/numpy/core/_methods.py", line 26, in _amax
    return umr_maximum(a, axis, None, out, keepdims)
ValueError: zero-size array to reduction operation maximum which has no identity
```
I'll look into it but wanted to see if you understand it off the top of your head.

milenaveneziani commented:
Mine goes, but for some reason it is only loading in 7 years of data, despite the fact that I have chosen year 2=9999 (there should be a total of 50 years for beta1). Let me look into it.

milenaveneziani commented on config.default:
```diff
@@ -132,6 +132,15 @@ mpasInterpolationMethod = bilinear
+#   like start_year = 1 and end_year = 9999 will be clipped to the valid range
+#   of years, and is a good way of insuring that all values are used.
+startYear = 2
+endYear = 6
```
oh, that's why. This must have ended up here by mistake.

milenaveneziani replied:
I think you need to remove this whole section starting with [index] from config.default.

vanroekel replied:
I added this in response to a request from @xylar to have an index category of analysis.

milenaveneziani replied:
I see. If we do that, then I would suggest to have startYear,endYear=1,9999 in the default and then add the same section in each example config file, because this is likely something that the user will change often.

xylar replied:
Yes, I agree, these need to be the defaults everywhere.

milenaveneziani replied:
I would also add a comment of the kind 'climate index time series should be longer than say 30 years to be meanginful'.

vanroekel replied:
Is there a reason to retain so many config files for each system?  I have updated the newest ones in each directory.  do I need to do all?  If so, I'd like to understand the need for all of them (for example configs/edison has 4 config files)

milenaveneziani replied:
We should have 3 on edison: one for an alpha run, one for beta0 and one for beta1. What's is the fourth one you see?

xylar replied:
The reason we want to keep all of these is that it's vitally important to test the analysis code constantly, on a number of machines, and using a number of runs we want to support.  We have tried to keep a minimum of relevant options in each that a user might be likely to change.

I use a different workflow myself where I feed 3 config files to `run_analysis.py`.  `config.default` happens automatically.  Then I feed in one of the machine config files, and finally a I pass in a config file that sets the output directory, and any options I'm likely to change from both the defaults and the specific run.  If everyone were doing that, we wouldn't need to include some of these in the machine-specific examples.  But I think my workflow is too complicated to become the default.

As always, we're open to suggestions for how to make things smoother for both users and developers.

vanroekel replied:
These files appear to be only functional with acme pre/post?  Some of them are very different from config.default (looking in the edison directory).  Should these be updated to be consistent with config.default?  I want to know what I am expected to change for this PR.

milenaveneziani replied:
The idea is to keep everything in config.defaults, but then re-define the options that the user is likely to have to modify for the specific run in the configs/$machine examples. Things like file locations, of course, but also start and end year for the various sections, directory paths, etc.
This is something that @xylar introduced recently and it turns out to be a time saver because before we had to update all the config examples, even when making changes to the config file that the user was unlikely to have to change routinely.

vanroekel replied:
I still very confused about what you want me to change.  For example in configs/edison, 
```
config.20161006bugfix.alpha8.A_WCYCL1850S.ne30_oEC_ICG.edison
```
Is missing most of the options in config.default.  Are you expecting me to add everything in there while respecting run directories and observational paths?

vanroekel replied:
Also, while I see the advantage of multiple config files for different machines.  I don't see the benefit of multiple files per machine.  Each file essentially only changes the run path, plot directory, and runName.  Doesn't seem like significant variability.

vanroekel replied:
I talked with @milenaveneziani a bit more about this and I now understand the benefit of the files.  I was not aware you could specify one of those and then if something is missing it is filled from config.default.  Definitely makes sense now.  I'm making the changes now.

vanroekel commented:
@xylar  This is related to me searching for the maximum and minimum y value for the plot range.  It doesn't work for short time series (maxIndex and minIndex in plotting.py need to be flipped for short time series).  I will ponder a more robust solution, but I don't think anyone should use this routine with fewer than ~30 years of data.

xylar commented:
I wasn't setting [index] yearStart and yearEnd so it was defaulting to values that don't produce a valid plot.  I have added them to my config file and now it seems to work.

I'm afraid you will have to add [index] and yearStart and yearEnd to all the config files under `configs`, since that might be an option a user is likely to want to change.

vanroekel commented:
sure, that's easy enough to add.

xylar commented:
I'm now getting the time series to plot just fine but the spectrum doesn't seem to have selected a reasonable axis.  This may just be a problem with having a ridiculously short time series.  I'll try on Edison after the ACME meeting.

![nino34_gmpas-qu240](https://cloud.githubusercontent.com/assets/4179064/24207410/09eb501a-0f21-11e7-9a65-1e4d02f537d5.png)
![nino34_spectra_gmpas-qu240](https://cloud.githubusercontent.com/assets/4179064/24207409/09c9a3b6-0f21-11e7-9fbc-5bfae85897bd.png)

xylar commented:
Beta0 on Edison looks good to me:
![nino34_20161117 beta0 a_wcycl1850 ne30_oec edison](https://cloud.githubusercontent.com/assets/4179064/24208368/c15db42a-0f23-11e7-87e8-e46f97d17795.png)
![nino34_spectra_20161117 beta0 a_wcycl1850 ne30_oec edison](https://cloud.githubusercontent.com/assets/4179064/24208367/c15caada-0f23-11e7-8f2e-bde5d0d6980e.png)

xylar commented:
Beta1 on Edison also looks good:
![nino34_20170313 beta1 a_wcycl1850s ne30_oecv3_icg edison](https://cloud.githubusercontent.com/assets/4179064/24209371/5566065c-0f26-11e7-9e51-793dd1cb1bf2.png)
![nino34_spectra_20170313 beta1 a_wcycl1850s ne30_oecv3_icg edison](https://cloud.githubusercontent.com/assets/4179064/24209372/55817f18-0f26-11e7-9590-cae655224f18.png)

xylar commented on mpas_analysis/ocean/nino34_index.py:
```diff
@@ -0,0 +1,226 @@
+    date_to_days, days_to_datetime
+
+# {{{
+def nino34_index(config, streamMap=None, variableMap=None):
```
@vanroekel, here and elsewhere, the opening fold `# {{{` should go after the function definition (after the colon). This way, when someone closes the fold, they see the name and arguments of the function but not the contents.  The end fold goes either at the end of the last line or on its own line below the last, depending on how long the last line is.

vanroekel replied:
sure, done.

xylar commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -27,6 +28,192 @@
+    ax.plot(f[2:-3], confidence99[2:-3], 'g', linewidth=linewidths)
+    ax.set_xlim(15, 1)
+
+#   add legend
```
As part of cleaning up formatting, could you align comments with the code above/below them as opposed to having the # at the beginning of the line?  I think PEP8 might complain about your formatting but even if not, this is the style we are using everywhere else.

vanroekel replied:
Just to note, the # at the beginning was because pep8 complained about # aligned with the code.

xylar replied:
Perhaps the issue was that there needs to be a space after each #?  I definitely align all my comments with the code.

xylar replied:
By the way, I know PEP8 seems super pedantic.  But it's helpful to have a strong standard to follow and it's the easiest one for us to automatically check for.

vanroekel replied:
I will certainly do this to follow the standards here, but I personally do not see the advantage of these pedantic formatting changes.  Again, my personal opinion.

xylar replied:
I fought against it at first myself.  But now that I'm using an editor that flags formatting as I go, I can really appreciate how much cleaner code can be when it's formatted uniformly as opposed to each file (or each routine) having a different style.  @pwolfram may have more to say, since he was the one who pushed hard for a uniform style.

milenaveneziani commented:
My beta1 test looks identical to Xylar's. I will now test on LANL IC, and then we should be done with testing, I think.

milenaveneziani commented:
all went well on wolf as well, even with a very short time series (I guess except for the x-axis issue noted already).

vanroekel commented:
I think I have finished all the formatting and clean-up.  let me know if more is necessary.

milenaveneziani commented:
thanks @vanroekel. I will do one last test turning on other time series as well, as you suggested.
Do you want to squash commits before merging?

vanroekel commented:
@milenaveneziani I'm ambivalent on squashing commits.  If you'd like me to, I'm happy to do so.  Would you like this to be one commit only?

milenaveneziani commented:
not important. 

@xylar: do you approve the changes?

xylar commented:
I know it's kind of a pain but I'd really like to have the commits squashed down to only those that are meaningful changes, not fixes or reformatting or things like that.  If it's easiest to just squash to a single commit, I'm good with that.

vanroekel commented:
that's fine, can do.  I'll look at the log and see what makes sense.

On Wed, Mar 22, 2017 at 4:01 PM, Xylar Asay-Davis <notifications@github.com>
wrote:

> I know it's kind of a pain but I'd really like to have the commits
> squashed down to only those that are meaningful changes, not fixes or
> reformatting or things like that. If it's easiest to just squash to a
> single commit, I'm good with that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MPAS-Dev/MPAS-Analysis/pull/146#issuecomment-288553604>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANB5UezHwCEVNHgPfJ_qVO2oAdAB2s-4ks5roZpDgaJpZM4MkDce>
> .
>

milenaveneziani commented:
just a minute: I found one feature missing in plot_timeseries..

milenaveneziani commented on mpas_analysis/shared/plot/plotting.py:
```diff
@@ -98,39 +298,7 @@ def timeseries_analysis_plot(config, dsvalues, N, title, xlabel, ylabel,
-    if yaxLimits[0]*yaxLimits[1] < 0:
-        indgood = np.where(np.logical_not(np.isnan(mean)))
-        x = mean['Time'][indgood]
-        plt.plot(x, np.zeros(np.size(x)), 'k-', linewidth=1.2)
```
the 6 lines above are now missing. Not sure where they should go. Perhaps before calling `_plot_xtick_format`?
They print a y=0 line when the y-axis spans negative and positive values, and it seemed useful for reading OHC time series in particular.

vanroekel replied:
sorry about that, let me reinstate and add squash commits.

vanroekel replied:
I think I have fixed this now.  Please test and I will then squash commits

milenaveneziani commented:
@vanroekel: all looks good now. I will merge after the squashing. Thanks.

vanroekel commented:
Commits are squashed.  Let me know if anything else is needed.

milenaveneziani commented:
Thanks for doing this @vanroekel.

vanroekel commented:
thanks all for the review and testing!

xylar commented:
Yes, thanks @vanroekel, for some really nice work and for putting up with the review process.  I know from my own experience that it can get really frustrating.  But the results are good, though.
